### PR TITLE
Add support for :raw flag to api and GET, etc

### PR DESCRIPTION
**Pull request by amalloy**
:raw true causes GET and friends to not destructure, coerce, or otherwise
fiddle with the request, except as specified in the compojure
request-destructuring clause - things like :query-params produce swagger
metadata, but do *not* cause any parameters to be destructured out of the
request.

Additionally, (api {:raw true ...} handler) support is added, which prevents
wrapping handler in the compojure-api middlewares.

Proposed implementation of the feature(s) requested in #240.

I wasn't sure how to add tests for this, since it's supposed to be hard to detect the features that compojure-api is adding: either your program depends on them, in which case removing them will lead to a compilation error, or it doesn't, in which case removing them will invisibly make things faster.

### CODE REVIEW

**Comment by amalloy** [comment]
As mentioned in #240, the `:raw true` flag for `GET` is unnecessary. I still like the idea of adding this flag to `api`, but it seems my implementation of that is not correct: when I add `:raw true` to the options I pass to `api`, it seems my `:produces` and `:consumes` annotations get ignored also, and I can't really figure out why that is. How does my `:raw true` flag change the `swagger.json` file that gets produced?

**Comment by ikitommi** [comment]
I have an idea to resolve these all, will check all these today evening (european). On vacation, not near computer now..

**Comment by ikitommi** [comment]
The `:produces` and `:consumes` are automatically set on the `api-middleware` based on the given ring-middleware-formats (and injected into request). If the `api-middleware` is not used, one has to set those manually:

```clj
(api
  {:swagger {:data {:produces ["application/json" "application/edn"]
                    :consumes ["application/json" "application/edn"]}}})
```

e.g. the `swagger-routes` read information from several sources and merges those together, see https://github.com/metosin/compojure-api/blob/master/src/compojure/api/swagger.clj#L43-L47 for sources & order of merging.

I think having an option not to mount the api-middleware is a good one. I think it could be renamed into something like `[:api :disable-api-middleware?]` - there is already one option under `:api`. Keeping the top-level options free of dirty details or something ;)

btw. One can now add `swagger-routes` to a vanilla compojure-app, just have to construct the whole (ring-)swagger-object by hand.

File: src/compojure/api/api.clj [hunk 1]
```diff
@@ -50,14 +52,19 @@
   [& body]
   (let [[options handlers] (common/extract-parameters body false)
         options (rsc/deep-merge api-defaults options)
-        handler (apply c/routes (concat [(swagger/swagger-routes (:swagger options))] handlers))
+        raw (:raw options)
+        swagger-ui (-> (swagger/swagger-routes (:swagger options))
+                       (cond-> raw ;; if the api won't be wrapped, swagger-ui still needs to be
```
Comment (ikitommi) [inline src/compojure/api/api.clj]: why does it have to be wrapped? feels odd if the swagger routes would have different format-settings.

Comment (amalloy) [inline src/compojure/api/api.clj]: Because you need that middleware to render the JSON into a string.

Comment (ikitommi) [inline src/compojure/api/api.clj]: but in the case you disable the whole `api-middleware`, you should have your own JSON-middleware setup outside of the `api` anyway to get any data out? Also, if you set your formats to `nil` and disable the `api-middleware`, this wrapping doesn't do the serialization...

Comment (ikitommi) [inline src/compojure/api/api.clj]: ... also, if you would do a EDN-only api for example, you could use the `swagger-routes` separately and wrap that with `wrap-json` / whatever to get JSON serialization just for that route.

Comment (amalloy) [inline src/compojure/api/api.clj]: For the first question, I'm not sure you should make assumptions about what middlewares the user has for their own routes; if there are middlewares needed for the swagger routes, those should be added for those routes specifically. For example, we don't have a json middleware that automatically detects clojure maps and converts them to json; instead we are explicit about json-encoding only the limited subset of json that we actually return from the server.

The second statement I don't understand. What do you mean, doesn't do the serialization? When I was writing this PR, I had formats set to `[:json]` or something like that, so it wasn't nil.

Comment (amalloy) [inline src/compojure/api/api.clj]: Yes, I mean now that you've told me there's a way to get just the swagger routes separately, without messing with the api middlewares, I think everything is just simpler. It was harder to add this feature when I thought I had to work it into `api`.

**Comment by ikitommi** [comment]
ok, thanks for explaining. Now I think I know your situation :) So, would it be a good idea just to:

- add an option `[:api :disable-api-middleware?]` that disables api-mw for all + use swagger-routes separately with a json-mw?

I would be happy with that, could cut out the 1.1.5 out.

**Comment by amalloy** [comment]
That would be fine I think, yes.

**Comment by ikitommi** [comment]
I was waiting for a commit with cleanup/rename, but guess miscommunicated my intention. I'll add this now.

**Comment by ikitommi** [comment]
Now in the latest `1.1.5-SNAPSHOT`.

**Pull request closed without merging**